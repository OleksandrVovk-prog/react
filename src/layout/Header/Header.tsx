import { Link } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useAppDispatch, useAppSelector } from '../../store/hooks/useApp';
import { toggleLocale } from '../../store/slices/translates/slice';
import { selectLocale } from '../../store/slices/translates/selectors';
import Title from '../../components/Title/Title';
import { selectUserId } from '../../store/slices/auth/selectors';
import authApi, { useFetchUserQuery } from '../../store/slices/auth/apis/auth';

import styles from './sass/Header.module.scss';

function Header(): JSX.Element {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const id = useAppSelector(selectUserId);
  const locale = useAppSelector(selectLocale);
  const { data } = useFetchUserQuery(id, { skip: !id });
  return (
    <header className={styles.header}>
      <Title text={t('common.welcomeMessage')} />
      <nav className={styles.headerNav}>
        <Link to="/" className={styles.headerNavLink}>
          { t('home.pageTitle') }
        </Link>
        <Link to="/about" className={styles.headerNavLink}>
          { t('about.pageTitle') }
        </Link>
      </nav>
      <div className={styles.toggleLanguage}>
        {data?.id ? (
          <Link
            to="/login"
            onClick={() => {
              dispatch(authApi.util.resetApiState());
            }}
            className={styles.headerNavLink}
          >
            { `${t('login.logout')}(${data.firstName})` }
          </Link>
        ) : (
          <Link
            to="/login"
            className={styles.headerNavLink}
          >
            { t('login.pageTitle') }
          </Link>
        )}
        <button
          data-testid="toggle-language-button"
          type="button"
          onClick={() => dispatch(toggleLocale())}
        >
          {locale.toUpperCase()}
        </button>
      </div>
    </header>
  );
}

export default Header;
