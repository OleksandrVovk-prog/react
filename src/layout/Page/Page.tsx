import { Outlet } from 'react-router-dom';
import Header from '../Header/Header';

function Page(): JSX.Element {
  return (
    <div className="page">
      <Header />
      <Outlet />
    </div>
  );
}

export default Page;
