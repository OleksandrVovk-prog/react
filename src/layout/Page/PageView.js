import { Outlet } from 'react-router-dom';
import Header from '../Header/Header';
import styles from './sass/Page.module.scss';

export default function PageView() {
  return (
    <div className={styles.app}>
      <Header />
      <Outlet />
    </div>
  );
}
