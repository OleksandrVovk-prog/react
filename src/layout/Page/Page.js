import { useCallback, useMemo } from 'react';
import i18n, { appLocales } from '../../i18n';
import PageView from './PageView';
import {
  Context as PageContext,
  data
} from '../../context/layout/Page';

export default function Page() {
  const { i18nConfig } = data;

  const changeLanguage = useCallback(() => {
    const nextLocaleIndex = appLocales.indexOf(i18n.language) + 1;
    const locale = nextLocaleIndex >= appLocales.length ? appLocales[0] : appLocales[nextLocaleIndex];
    i18n.changeLanguage(locale);
  }, []);

  const pageContextValue = useMemo(() => ({ changeLanguage, i18nConfig }), [i18nConfig, changeLanguage]);

  return (
    <PageContext.Provider value={pageContextValue}>
      <PageView />
    </PageContext.Provider>
  );
}
