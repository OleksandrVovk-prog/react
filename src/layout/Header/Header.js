import { Link } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import Title from '../../components/Title/Title';
import { Context as PageContext } from '../../context/layout/Page';
import styles from './sass/Header.module.scss';

export default function Header() {
  const { i18n, t } = useTranslation();

  return (
    <PageContext.Consumer>
      {(config) => (
        <header className={styles.appHeader}>
          <Title text={t('welcome')} />
          <Link to="/" className={styles.headerLink}>
            {t('home')}
          </Link>
          <Link to="/about" className={styles.headerLink}>
            {t('about')}
          </Link>
          <Link to="/lazy" className={styles.headerLink}>
            {t('lazy')}
          </Link>
          <Link to="/memo" className={styles.headerLink}>
            {t('memo')}
          </Link>
          <Link to="/context" className={styles.headerLink}>
            {t('context')}
          </Link>
          <button
            className={styles.toggleLanguage}
            type="button"
            onClick={config.changeLanguage}
            data-testid="toggle-language-button"
          >
            {i18n.language}
          </button>
        </header>
      )}
    </PageContext.Consumer>
  );
}
