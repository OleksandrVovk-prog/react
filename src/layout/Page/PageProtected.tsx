import { Navigate } from 'react-router-dom';
import { useAppSelector } from '../../store/hooks/useApp';
import { selectUserId } from '../../store/slices/auth/selectors';
import Page from './Page';

import IPageProtected from './interfaces/IPageProtected';

/**
 * Protected page layout
 */
function PageProtected({ redirectUri }: IPageProtected): JSX.Element {
  const id = useAppSelector(selectUserId);
  return id ? (
    <Page />
  ) : (
    <Navigate to={redirectUri} />
  );
}

export default PageProtected;
