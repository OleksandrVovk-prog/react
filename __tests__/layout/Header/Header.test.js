import React from 'react';
import { render } from '../../settings/test-utils';
import Header from '../../../src/layout/Header/Header';
import { Context as PageContext } from '../../../src/context/layout/Page';

describe('<Header />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = render(
      <PageContext.Consumer>
        { () => <Header /> }
      </PageContext.Consumer>
    );
  });
  test('should render Header component', () => {
    const { asFragment } = wrapper;
    expect(asFragment(
      <Header />
    )).toMatchSnapshot();
  });
});
