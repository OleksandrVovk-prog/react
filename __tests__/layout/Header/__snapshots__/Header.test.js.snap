// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render Header component 1`] = `
<DocumentFragment>
  <header
    class="appHeader"
  >
    <h1
      class="appTitle"
    >
      Try to build something great
    </h1>
    <a
      class="headerLink"
      href="/"
    >
      Home
    </a>
    <a
      class="headerLink"
      href="/about"
    >
      About
    </a>
    <a
      class="headerLink"
      href="/lazy"
    >
      Lazy
    </a>
    <a
      class="headerLink"
      href="/memo"
    >
      Memo
    </a>
    <a
      class="headerLink"
      href="/context"
    >
      Context
    </a>
    <button
      class="toggleLanguage"
      data-testid="toggle-language-button"
      type="button"
    >
      en
    </button>
  </header>
</DocumentFragment>
`;
